#303 - feat(feedback): Resizable MapMetaData Window

Kadestery opened the pull request:
### Description

This PR updates the frontend, the MapMetaData window is now resizable.

---

### Frontend Implementation

- **`MapMetaData`**: Added the ability to resize the windows' width, it has a minimum(300px) and a maximum(1/3rd of screen) width. Also made the window responsive to adapt to smaller screen sizes.

---

### Testing

- Unit tests for MapMetaData resizing component `(Resizing Behaviour)`

---

### Acceptance Criteria Addressed

- The MapMetaData initializes with default 400px width 
- The MapMetaData window maintains width when toggling collapse
- The MapMetaData window respects the minimum and maximum width constraints
- The MapMetaData window resizes via the handle

---

### Screenshots
**(Menu when default MapMetaData width)**:
![image](https://github.com/user-attachments/assets/80d73683-3aeb-47c0-8d72-849f92df0c4b)
**(Menu when maximum MapMetaData width)**:
![image](https://github.com/user-attachments/assets/6835d09f-dade-41d2-8651-a08d1f9cff09)

---

Substantial Engineering has been elaborated below:
https://github.com/XavierGuertin/WildfireVisualizationProject/issues/303#issuecomment-2751544707

Kadestery committed:
#303 - Modified README

Added PowerShell option for Developers when adding environment variables

Kadestery committed:
#303 - Modified files in order to function

Added @Getter to DTO files to build getter methods, and refactored pom.xml dependencies

Kadestery committed:
#303 - Modified files to function p2

Added toastify dependency, changed geoserver image release, and port changed from 8090:8080 to 8091:8080. Disabled Figlet Banner, was causing issue.

Kadestery committed:
#303 - Resizing MapMetaData implementation complete

Added a resizing width effect using users' mouse, and a dynamic icon. Max width of 1/3 of the screen. Also, made the Load Dataset button and Weather Assets dropdown button the same width throughout resizing

Kadestery committed:
#303 - Added tests for resizing mapmetadata

Modified for MapMetaData test file to have nesting, looks cleaner. Added min. width test, max. width test, maintains width when toggling...

Kadestery committed:
#303 - Modified resizing animation

Made resizing to have 0 delay, smoother for user

Kadestery committed:
#303 - Modify files to pass frontend CI

Added "es lint." script to package.json, in order to pass CI using lint. Fixed 4 errors, and 6 warnings.

Kadestery committed:
#303 - Fixed MapMetaData tests

Miss input of the initial width

Kadestery committed:
#303 - Fixed MapMetaData tests

Modified tests to measure the window as opposed to using a known width

Kadestery committed:
#303 - Modified MapMetaData tests

Performed the mouseUp prior to, and expect of container being 0 is false, the container is there, just transparent

Kadestery committed:
#303 - Finalizing MapMetaData Tests

Modified the "resize via handle" test, incorporating the offset of viewport for mouseDown

Kadestery committed:
#303 - Final MapMetaData Test Change

Had to change the arithmetic for the viewport offset

Kadestery commented:
### 1. Understanding the Requirements
The task was to improve the visual representation of MapMetaData information. Previously, the window was only 400 pixels wide, hence when there were many sentences in the description of a dataset, a lot would be hidden and need scrolling to view. We wanted users to have greater readability. Our goal was to enhance this area by:

- Allowing the metadata window to be resizable by the user.
- Enforce a maximum width constraint of 1/3 of the screen to ensure it does not obstruct too much of the map view or UI.
- Make the window responsive so that it adapts to smaller screen sizes.

We clarified these requirements by reviewing the existing UI and discussing them with the stakeholders.

---

### 2. Frontend & UX Enhancements
On the frontend, a key modification was introduced:

- **Resizable MapMetaData window with Icon:**  
  The MapMetaData window now has a slider icon, so users can easily understand its' intent. When the window approaches its' minimum width, the descriptions have scrollable text boxes. We made it a minimum width of 350 pixels and a maximum of 1/3rd the width of the screen, according to our stakeholder's needs.


- Additional UI Enhancements:

   - Updated the width of the "Weather Assets" dropdown button to be the same width as the "Load Dataset" button

---

### 3. Technical Challenges
- **Errors with Running:**
The program required getter methods from Dto classes. Went back and forth between implementing them manually and using Lombok. Decided to use Lombok and added "@Getter" to the classes, as this aided in the readability of the code. Geoservers' port was also being used, had to change it from 8090 to 8091. There was a missing FIGlet file, had to disable it and also set the image for Kartoza/Geoserver to '2.23.1' instead of 'latest'.

- **CI/CD Pipeline & Testing:**
The Frontend CI initially failed due to missing linting. Adding 'eslint .' resolved this, but intermittent backend failures persisted. By refactoring tests using nested structures for clarity and fixing slider offset calculations via getBoundingClientRect(), we achieved 95%+ pass rates.

- **Performance Optimization:**
Smooth resizing required disabling CSS animations and setting transition: 0s. This reduced lag by avoiding unnecessary repaints, validated via Chrome DevTools’ performance profiler.

### 4. Manual Testing
Manual testing was performed by @Kadestery:

- The window resizes smoothly.

- Maximum width is good for what the stakeholder wanted... can still see the majority of the map with a larger description view

### 5. Final Thoughts / Engineering Outcome
The implemented solution successfully enhances the MapMetaData windows' usability/accessibility. 
Key benefits include:

- **Enhanced Usability:**
Users are now able to see the whole description without needing to scroll and see the map at the same time.


Overall, the changes create a more usable interface that aligns with our commitment to continuous UI improvement and modern design principles.

XavierGuertin commented on apps/backend/pom.xml:
```diff
@@ -53,44 +37,61 @@
     <dependency>
-      <groupId>org.springframework</groupId>
-      <artifactId>spring-jdbc</artifactId>
+      <groupId>org.springframework.boot</groupId>
+      <artifactId>spring-boot-starter-web</artifactId>
+      <exclusions>
+        <exclusion>
+          <groupId>org.springframework.boot</groupId>
+          <artifactId>spring-boot-starter-logging</artifactId>
+        </exclusion>
+      </exclusions>
+    </dependency>
+
+    <dependency>
+      <groupId>org.springframework.boot</groupId>
+      <artifactId>spring-boot-starter-jdbc</artifactId>
     </dependency>
 
     <dependency>
       <groupId>org.springframework.boot</groupId>
       <artifactId>spring-boot-starter-actuator</artifactId>
     </dependency>
 
+    <dependency>
+      <groupId>org.springframework.boot</groupId>
+      <artifactId>spring-boot-starter-log4j2</artifactId>
+    </dependency>
+
+    <!-- Database -->
+    <dependency>
+      <groupId>org.postgresql</groupId>
+      <artifactId>postgresql</artifactId>
+      <version>42.7.5</version>
+    </dependency>
```
Why did you add all of these dependencies. Same applies for the other ones below

Kadestery replied:
I didn't add any new dependencies, I only reformatted the existing code for better readability/organization. The dependency structure remains unchanged from the original

Kadestery replied:
Returning the pom.xml file as it was, due to being out of the scope for my issue.

XavierGuertin commented on apps/frontend/package.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "dependencies": {
+    "react-toastify": "^11.0.5"
+  },
+  "scripts": {
+    "lint": "eslint ."
```
why? We already have toast defined.

Kadestery replied:
I noticed toast was already defined, but I had to add toastify to package.json for the program to run locally on my machine. I’ll investigate why this dependency was needed and check for potential conflicts/possibilities of removing it.

lo-vil replied:
This lack of dependency was causing issue when you ran nx dev frontend ?

Kadestery replied:
Removed "react-toastify" as it was an error on my part for keeping it included.

XavierGuertin commented on docker-compose.yml:
```diff
@@ -44,7 +44,7 @@ services:
     environment:
       - NEXT_PUBLIC_BACKEND_URL=http://localhost:8080
-      - NEXT_PUBLIC_GEOSERVER_URL=http://localhost:8090/geoserver/Default/wms
+      - NEXT_PUBLIC_GEOSERVER_URL=http://localhost:8091/geoserver/Default/wms
```
why

Kadestery replied:
As mentioned in the 'Technical Challenges - Errors with Running' section of https://github.com/XavierGuertin/WildfireVisualizationProject/issues/303#issuecomment-2751544707, there was a port conflict (8090 already in use), so I switched to 8091 for GeoServer.

For testing, I can temporarily use 8091 locally and revert to 8090 before pushing if needed. Let me know your preference.

lo-vil replied:
I don't think this should be pushed before discussing with the team. All of workflows and env variables reference port 8090, this change has a lot of consequences

Kadestery replied:
Talked with Sam and was recommended to keep on 8090. Will revert to the 8090 port.

XavierGuertin commented on docker-compose.yml:
```diff
@@ -124,14 +124,16 @@ services:
-    image: kartoza/geoserver:latest
+    image: kartoza/geoserver:2.23.1
     ports:
-      - "8090:8080"
+      - "8091:8080"
     environment:
       - GEOSERVER_ADMIN_USER=geoserver
       - GEOSERVER_ADMIN_PASSWORD=password
       - GEOSERVER_DATA_DIR=/var/geoserver/data_dir
-      - RUN_AS_ROOT=TRUE
+      - INITIAL_MEMORY=2G
+      - MAXIMUM_MEMORY=4G
+      - DISABLE_FIGLET_BANNER=true  # Explicitly disable figlet
```
why

Kadestery replied:
As mentioned in the 'Technical Challenges - Errors with Running' section of https://github.com/XavierGuertin/WildfireVisualizationProject/issues/303#issuecomment-2751544707, there was a port conflict (8090 already in use), so I switched to 8091 for GeoServer.

For testing, I can temporarily use 8091 locally and revert to 8090 before pushing if needed. Let me know your preference.

lo-vil replied:
Same comment as above. This is also out of scope for this task

Kadestery replied:
Talked with Sam and was recommended to keep on 8090. Will revert to the 8090 port.

Kadestery committed:
#303 - Reverse Modification of Out-scope File

In order to run the program on my end, needed to change some files that were out of my scope for the issue I was assigned. Needed to reverse the changes made in pom.xml, package.json (toastify), and docker-compose.yml (8090 port).

Kadestery committed:
Merge branch 'dev' into 303/width-resize-metadata

Kadestery committed:
#303 - Fix Lint Frontend

Error with init being empty in reactI18next.js

lo-vil commented on apps/frontend/package.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "scripts": {
+    "lint": "eslint ."
+  }
+}
```
I don't think we need this, i believe all the dependencies are handled at the root package.json by nx

Kadestery replied:
At first, in order for the frontend tests to pass/be-tested I needed to add that to the file, I'll remove it and try again!

Kadestery replied:
![image](https://github.com/user-attachments/assets/65c09bf7-343c-4aa5-8086-bfd75421302b)

When I push to GitHub, this is what I get when I remove "eslint ." from package.json.

lo-vil replied:
That's weird indeed. Could it have to do with these though: 

<img width="932" alt="Screenshot 2025-03-27 at 11 46 21" src="https://github.com/user-attachments/assets/dd8b086b-d0ae-4856-8eaa-48cd9c837b0f" />

We have these ignore rules already in the eslint.config.ts at the root:

<img width="1184" alt="Screenshot 2025-03-27 at 11 50 28" src="https://github.com/user-attachments/assets/1874abda-3aab-48af-ac72-e87f8a7af5cd" />

Just trying to understand what's going on, thanks for the quick response

Kadestery replied:
Thanks for helping! I'll remove that too! I think it may have to do with the tests I implemented and not passing the TypeScript linting. I'm going to take a look at the tests!

Kadestery committed:
Revert "#303 - Modified README"

This reverts commit c391098ab1eda09831274f2398e155b4a6a92e10.

Kadestery committed:
#303 - Fully Reverted Out-Scope Files

Removed @Getter from DTO files, removed eslint from package.json, and removed modifications to ReadMe file

Kadestery committed:
#303 - Modified MapMetaData Tests for proper Linting

Ensured all async operations use 'act', properly typed mouse events.

Kadestery committed:
Revert "#303 - Modified MapMetaData Tests for proper Linting"

This reverts commit c304aeeb5beffd854e897401ffe8d1fd95668289.

Kadestery committed:
#303 - Removed package.json in frontend

By accident have been using frontend package.json to run linting, even though it's also in root. Deleted the file, now linting runs

Kadestery committed:
Merge branch 'dev' into 303/width-resize-metadata

lo-vil reviewed (approved):
Nice job man ! Env was rough but great perseverance 💯

im-samwong reviewed (approved):
good.
